features/FEATURE_BLE/source/generic/GenericGap.cpp

```diff
@@ -0,0 +1,1091 @@
+        handle,
+        params->minConnectionInterval,
+        params->maxConnectionInterval,
+        params->slaveLatency,
+        params->connectionSupervisionTimeout,
+        /* minimum_connection_event_length */ 0,
+        /* maximum_connection_event_length */ 0
+    );
+}
+
+ble_error_t GenericGap::getPreferredConnectionParams(ConnectionParams_t *params)
+{
+    if (params == NULL) {
+        return BLE_ERROR_INVALID_PARAM;
+    }
+
+    return _gap_service.get_peripheral_prefered_connection_parameters(
+        *params
+    );
+}
+
+ble_error_t GenericGap::setPreferredConnectionParams(const ConnectionParams_t *params)
+{
+    if(is_preferred_connection_params_valid(params) == false) {
+        return BLE_ERROR_PARAM_OUT_OF_RANGE;
+    }
+
+    return _gap_service.set_peripheral_prefered_connection_parameters(
+        *params
+    );
+}
+
+ble_error_t GenericGap::setDeviceName(const uint8_t *deviceName)
+{
+    return _gap_service.set_device_name(deviceName);
+}
+
+ble_error_t GenericGap::getDeviceName(uint8_t *deviceName, unsigned *lengthP)
+{
+    if (lengthP == NULL) {
+        return BLE_ERROR_INVALID_PARAM;
+    }
+
+    uint8_t length = 0;
+    ble_error_t err = _gap_service.get_device_name_length(length);
+    if (err) {
+        return err;
+    }
+
+    if (deviceName != NULL) {
+        if (*lengthP < length) {
+            return BLE_ERROR_INVALID_PARAM;
+        }
+
+        ArrayView<uint8_t> name(deviceName, *lengthP);
+        err = _gap_service.get_device_name(name);
+        if (err) {
+            return err;
+        }
+    }
+    *lengthP = length;
+    return BLE_ERROR_NONE;
+}
+
+ble_error_t GenericGap::setAppearance(GapAdvertisingData::Appearance appearance)
+{
+    return _gap_service.set_appearance(appearance);
+}
+
+ble_error_t GenericGap::getAppearance(GapAdvertisingData::Appearance *appearanceP)
+{
+    if (appearanceP == NULL) {
+        return BLE_ERROR_INVALID_PARAM;
+    }
+    return _gap_service.get_appearance(*appearanceP);
+}
+
+ble_error_t GenericGap::setTxPower(int8_t txPower)
+{
+    // TODO: This is not standard, expose it as an extension API and document it
+    // as such
+    return BLE_ERROR_NOT_IMPLEMENTED;
+}
+
+void GenericGap::getPermittedTxPowerValues(const int8_t **valueArrayPP, size_t *countP)
+{
+    *countP = 0;
+}
+
+uint8_t GenericGap::getMaxWhitelistSize(void) const
+{
+    return _pal_gap.read_white_list_capacity();
+}
+
+ble_error_t GenericGap::getWhitelist(Whitelist_t &whitelist) const
+{
+    if(initialize_whitelist() == false) {
+        return BLE_ERROR_INVALID_STATE;
+    }
+
+    if (whitelist.capacity < _whitelist.capacity) {
+        return BLE_ERROR_INVALID_PARAM;
+    }
+
+    for (size_t i = 0; i < _whitelist.size; ++i) {
+        whitelist.addresses[i] = _whitelist.addresses[i];
+    }
+
+    whitelist.capacity = _whitelist.capacity;
+    return BLE_ERROR_NONE;
+}
+
+ble_error_t GenericGap::setWhitelist(const Whitelist_t &whitelist)
```

Comment:
I dont get why this actoion is so complicated - why it;s not about just clearing the whitelist and filing it again?

Reply:
The goal is to reduce HCI transactions; Clearing the whitelist and add a new one is simpler but not optimal.